mongoDB.cabal

```diff
@@ -54,14 +60,22 @@ Library
                     , base64-bytestring >= 1.0.0.1
                     , nonce >= 1.0.5
 
+  if flag(_old-network)
```

Comment:
@scott-fleischman Can you clarify for me who and how is supposed to set this flag?

Reply:
They are set using the cabal configuration flags as described [here](https://www.haskell.org/cabal/users-guide/developing-packages.html#configurations).

There are a couple choices involved here that I basically deferred to the choices made in the bson PR.
1. The name of the flag. `_old-network` is pretty cryptic. I just copied it. If flags can contain numbers it might be better to name it something like `use-network2` or something like that.
2. Default value. Again I made it `False` by default but maybe until `network-3.*` gets into stackage, maybe the default should be true? Then when stackage includes network 3 by default in the newest stackage lts then mongoDB could ship a breaking change which changes the flag to true, so it defaults to using network3.

I don't believe the `.cabal` files themselves allow logic like conditional statements based on a library dependency version. In our case we would need to include a new library (`network-bsd`) if network is `>=3.0`, so that might not be possible without configure flags.

In the end, I am happy to consider alternatives to the approach taken in this PR with regards to cabal configuration. But as far as I can tell that is going to involve some user choice of which network to use.

You can see how I set them in the stack files, for example:
https://github.com/plow-technologies/mongodb/blob/f84cc035179198ed63a5e247aa7c95f61ce7a295/stack-ghc80.yaml#L2-L4

The flag is omitted here, so it defaults to using the network 3.0+ and network-bsd dependencies.
https://github.com/plow-technologies/mongodb/blob/network-deprecations/stack-ghc86-network3.yaml

Reply:
Maybe `manual: True` might be useful here though. https://www.haskell.org/cabal/users-guide/developing-packages.html#pkg-field-flag-manual

> By default, Cabal will first try to satisfy dependencies with the default flag value and then, if that is not possible, with the negated value. However, if the flag is manual, then the default value (which can be overridden by commandline flags) will be used.

Reply:
It doesn't seem to work for `stack` if I remove the explicit flag set. I get errors in the stack configure step.

`cabal` (cabal-install version 2.4.0.0) itself seems ok with it though. I was able to build using ghc 8.2.2 and it picked the old network configurations.

Reply:
So all in all, it seems to me the current approach seems pretty good given the various options I am aware of.

Stack users will have to set the flag explicitly for network <=2.8.
Cabal users should work fine as-is. (I'm not sure how to test against the newer network because I'm not sure how to get cabal to know about the bson PR, which it would need to succeed.)

So with defaulting the flag to false, stack users of network 2.8 will have to explicitly set it to true (or explicitly include network 3). When the stackage lts includes network 3, then can remove the explicit flag from their configuration.

Cabal users should work fine in both cases. When network 3 is available for their configuration (needs bson available that works with network 3), since the `_old-network` flag default to True, it will try the network 3 config first, so it will use that. Else it default to the old network configuration and builds fine. I tested with ghc 8.2.2 and ghc 8.6.5 with cabal 2.4.

Reply:
Here's a reference to the conditional options in a cabal file: https://www.haskell.org/cabal/users-guide/developing-packages.html#conditions